url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/Eval2.java

```diff
@@ -0,0 +1,1640 @@
+
+                if("set".equals(lLstExpr.get(0))) lCtx.setBinding(lNameRepr, lData[1]);
+                else lCtx.getRootContext().defBinding(lNameRepr, lData[1]);
+
+                aFrame.setResult(lData[1]);
+            }
+        }
+
+        public void init(Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+        throws CommandException
+        {
+        }
+    }
+
+    private static final Eval2.IFrameHandler PAIR_HANDLER = new PairHandler();
+    private static final Eval2.IFrameHandler COUPLE_HANDLER = new CoupleHandler();
+
+    public void init(Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        // Allocate enpough space to evaluate the left and right side of the assignment.
+        aFrame.allocateData(2);
+
+        final List lLstExpr = (List) aFrame.getExpr();
+        final int lLstSize = lLstExpr.size();
+
+        // Replace the handlers with more specific ones.
+        if(lLstSize == 2)
+        {
+            // Variant (xxx name=value)
+            aFrame.setHandler(PAIR_HANDLER);
+        }
+        else if(lLstSize == 3)
+        {
+            // Variant (xxx name value)
+            aFrame.setHandler(COUPLE_HANDLER);
+        }
+        else
+        {
+            throw new CommandException(String.format("The '%s' form should have the format (%s name value).",lLstExpr.get(0), lLstExpr.get(0)), aStack);
+        }
+    }
+
+    public void handleFrame(IEval aEval, Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        // This handler will never be called, by now the frame handler will be replaced
+        // by a more specific handler.
+    }
+}
+
+class LetHandler
+implements Eval2.IFrameHandler
+{
+    public void init(Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        final List lLstExpr = (List) aFrame.getExpr();
+        final int lLstSize = lLstExpr.size();
+        final IContext lCtx = aFrame.getCtx();
+
+        if(lLstSize != 3) throw new CommandException(String.format("The '%s' form should have the format (%s ((name val)...) expr).",lLstExpr.get(0), lLstExpr.get(0)), aStack);
+        final Object lBindings = lLstExpr.get(1);
+
+        // Check the type of the list of bindings.
+        if(!(lBindings instanceof List))
+            throw new CommandException(String.format("The '%s' form should have the format (%s ((name val)...) expr).\nThe first parameter should be a list of bindings but encountered an instance of type '%s'.",lLstExpr.get(0), lLstExpr.get(0), lBindings==null?"null":lBindings.getClass().getCanonicalName()), aStack);
+
+        // Allocate space for all new bindings and also the expression.
+        aFrame.allocateData(((List)lBindings).size() + 1);
+
+        // Install a new context.
+        aFrame.setCtx(new CompositeContext(new BasicContext(), lCtx));
+    }
+
+    public void handleFrame(IEval aEval, Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.